subsys/usb/device_next/class/usbd_cdc_acm.c

```diff
@@ -0,0 +1,1131 @@
+	net_buf_add(buf, len);
+
+	ret = usbd_ep_enqueue(c_nd, buf);
+	if (ret) {
+		LOG_ERR("Failed to enqueue");
+		net_buf_unref(buf);
+	}
+
+tx_fifo_handler_exit:
+	atomic_clear_bit(&data->state, CDC_ACM_LOCK);
+}
+
+/*
+ * RX handler should be conditionally triggered at:
+ *  - (x) cdc_acm_irq_rx_enable()
+ *  - (x) RX transfer completion
+ *  - (x) the end of cdc_acm_irq_cb_handler
+ *  - (x) USBD class API enable call
+ *  - ( ) USBD class API resumed call (TODO)
+ */
+static void cdc_acm_rx_fifo_handler(struct k_work *work)
+{
+	struct cdc_acm_uart_data *data;
+	struct usbd_class_node *c_nd;
+	struct net_buf *buf;
+	uint8_t ep;
+	int ret;
+
+	data = CONTAINER_OF(work, struct cdc_acm_uart_data, rx_fifo_work);
+	c_nd = data->c_nd;
+
+	if (!atomic_test_bit(&data->state, CDC_ACM_CLASS_ENABLED) ||
+	    atomic_test_bit(&data->state, CDC_ACM_CLASS_SUSPENDED)) {
+		LOG_INF("USB configuration is not enabled or suspended");
+		return;
+	}
+
+	if (ring_buf_space_get(data->rx_fifo.rb) < cdc_acm_get_bulk_mps(c_nd)) {
+		LOG_INF("RX buffer to small, throttle");
+		return;
+	}
+
+	if (atomic_test_and_set_bit(&data->state, CDC_ACM_RX_FIFO_BUSY)) {
+		LOG_WRN("RX transfer already in progress");
+		return;
+	}
+
+	ep = cdc_acm_get_bulk_out(c_nd);
+	buf = cdc_acm_buf_alloc(ep);
+	if (buf == NULL) {
+		return;
+	}
+
+	ret = usbd_ep_enqueue(c_nd, buf);
+	if (ret) {
+		LOG_ERR("Failed to enqueue net_buf for 0x%02x", ep);
+		net_buf_unref(buf);
+	}
+}
+
+static void cdc_acm_irq_tx_enable(const struct device *dev)
+{
+	struct cdc_acm_uart_data *const data = dev->data;
+
+	atomic_set_bit(&data->state, CDC_ACM_IRQ_TX_ENABLED);
+
+	if (ring_buf_is_empty(data->tx_fifo.rb)) {
+		LOG_INF("tx_en: trigger irq_cb_work");
+		cdc_acm_work_submit(&data->irq_cb_work);
+	}
+}
+
+static void cdc_acm_irq_tx_disable(const struct device *dev)
+{
+	struct cdc_acm_uart_data *const data = dev->data;
+
+	atomic_clear_bit(&data->state, CDC_ACM_IRQ_TX_ENABLED);
+}
+
+static void cdc_acm_irq_rx_enable(const struct device *dev)
+{
+	struct cdc_acm_uart_data *const data = dev->data;
+
+	atomic_set_bit(&data->state, CDC_ACM_IRQ_RX_ENABLED);
+
+	/* Permit buffer to be drained regardless of USB state */
+	if (!ring_buf_is_empty(data->rx_fifo.rb)) {
+		LOG_INF("rx_en: trigger irq_cb_work");
+		cdc_acm_work_submit(&data->irq_cb_work);
+	}
+
+	if (!atomic_test_bit(&data->state, CDC_ACM_RX_FIFO_BUSY)) {
+		LOG_INF("rx_en: trigger rx_fifo_work");
+		cdc_acm_work_submit(&data->rx_fifo_work);
+	}
+}
+
+static void cdc_acm_irq_rx_disable(const struct device *dev)
+{
+	struct cdc_acm_uart_data *const data = dev->data;
+
+	atomic_clear_bit(&data->state, CDC_ACM_IRQ_RX_ENABLED);
+}
+
+static int cdc_acm_fifo_fill(const struct device *dev,
+			     const uint8_t *const tx_data,
+			     const int len)
+{
+	struct cdc_acm_uart_data *const data = dev->data;
+	uint32_t done;
+
+	if (!check_wq_ctx(dev)) {
+		LOG_WRN("Invoked by inappropriate context");
+		__ASSERT(true, "Invoked by inappropriate context");
```

Comment:
Asserting on true makes no sense. Condition `true` is always true and therefore always is fulfilled.

The assert should be `__ASSERT(check_wq_ctx(dev), "Invoked by inappropriate context");` and be placed before the if statement.

Reply:
No, it should be `__ASSERT(false, "Invoked by inappropriate context");`, alternative is no ASSERT at all.